File: agent/tools/execution.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Literal

from loguru import logger

from ..control import TurnPaused
from ..providers import ToolCallRequest
from ..runtime import events as runtime_events
from .registry import ToolRegistry
from .results import ToolResult

StreamEmitter = Callable[[dict[str, Any]], Awaitable[None]]
ToolRunStatus = Literal["queued", "executing", "completed", "failed", "cancelled"]


@dataclass
class ToolRunState:
    id: str
    name: str
    arguments: dict[str, Any]
    status: ToolRunStatus = "queued"
    concurrency_safe: bool = False
    result: ToolResult | None = None
    error: str | None = None


class ToolExecutionEngine:
    def __init__(self, registry: ToolRegistry) -> None:
        self.registry = registry

    async def run_batch(
        self,
        tool_calls: list[ToolCallRequest],
        *,
        emit: StreamEmitter | None = None,
        run_one=None,
    ) -> list[dict[str, Any]]:
        states = [self._state_for_call(call) for call in tool_calls]
        if emit:
            for state in states:
                await emit(runtime_events.tool_run_queued(id=state.id, name=state.name, arguments=state.arguments))
        results_by_id: dict[str, ToolResult] = {}
        index = 0
        while index < len(tool_calls):
            state = states[index]
            if state.concurrency_safe:
                group_calls: list[ToolCallRequest] = []
                group_states: list[ToolRunState] = []
                while index < len(tool_calls) and states[index].concurrency_safe:
                    group_calls.append(tool_calls[index])
                    group_states.append(states[index])
                    index += 1
                gathered = await asyncio.gather(
                    *[
                        self._run_state(call, item, emit=emit, run_one=run_one)
                        for call, item in zip(group_calls, group_states, strict=True)
                    ],
                    return_exceptions=True,
                )
                for call, item, raw in zip(group_calls, group_states, gathered, strict=True):
                    if isinstance(raw, TurnPaused):
                        raise raw
                    if isinstance(raw, Exception):
                        result = ToolResult.from_text(f"Error: {raw}", is_error=True)
                        item.status = "failed"
                        item.error = str(raw)
                        results_by_id[call.id] = result
                        if emit:
                            await emit(runtime_events.tool_run_failed(id=call.id, name=call.name, message=str(raw)))
                    else:
                        results_by_id[call.id] = raw
                continue
            results_by_id[tool_calls[index].id] = await self._run_state(
                tool_calls[index],
                state,
                emit=emit,
                run_one=run_one,
            )
            index += 1
        return [
            {
                "role": "tool",
                "tool_call_id": call.id,
                "name": call.name,
                "content": results_by_id.get(call.id, ToolResult.from_text("")).model_content,
            }
            for call in tool_calls
        ]

    def _state_for_call(self, call: ToolCallRequest) -> ToolRunState:
        tool = self.registry.get(call.name)
        concurrency_safe = bool(tool is not None and tool.is_concurrency_safe(call.arguments))
        return ToolRunState(
            id=call.id,
            name=call.name,
            arguments=call.arguments,
            concurrency_safe=concurrency_safe,
        )

    async def _run_state(
        self,
        call: ToolCallRequest,
        state: ToolRunState,
        *,
        emit: StreamEmitter | None,
        run_one,
    ) -> ToolResult:
        state.status = "executing"
        if emit:
            await emit(runtime_events.tool_run_started(id=state.id, name=state.name))
        try:
            if run_one is None:
                result = await asyncio.to_thread(self.registry.execute_result, call.name, call.arguments)
            else:
                result = _coerce_tool_result(await run_one(call))
        except TurnPaused:
            state.status = "cancelled"
            if emit:
                await emit(runtime_events.tool_run_cancelled(id=state.id, name=state.name, reason="turn_paused"))
            raise
        except Exception as exc:
            logger.exception(f"Tool execution failed: {call.name}")
            state.status = "failed"
            state.error = str(exc)
            if emit:
                await emit(runtime_events.tool_run_failed(id=state.id, name=state.name, message=str(exc)))
            return ToolResult.from_text(f"Error: {exc}", is_error=True)
        state.status = "failed" if result.is_error else "completed"
        state.result = result
        if emit:
            if result.is_error:
                await emit(runtime_events.tool_run_failed(id=state.id, name=state.name, message=result.summary))
            else:
                await emit(runtime_events.tool_run_completed(
                    id=state.id,
                    name=state.name,
                    summary=result.summary,
                    artifacts=result.artifact_payloads() or None,
                    metadata=result.metadata or None,
                ))
        return result


def _coerce_tool_result(value: Any) -> ToolResult:
    if isinstance(value, ToolResult):
        return value
    text = str(value)
    return ToolResult.from_text(text, is_error=text.startswith("Error:"))

File: packages/core/src/tools/execution.ts
/**
 * ToolExecutionEngine (MIG-CORE-008 支撑)。对齐 Python `agent/tools/execution.py`。
 * concurrency_safe 连续工具成组并发，其余顺序执行；emit tool_run_* 事件；TurnPaused 冒泡。
 */
import type { ToolCallRequest } from '../providers/base'
import { ToolResultObj } from './base'
import type { ToolRegistry } from './registry'
import { TurnPaused } from '../control/exceptions'
import * as runtimeEvents from '../agent/runtime-events'
import { CancelledTaskError } from '../runtime/active'

export type StreamEmitter = (event: Record<string, unknown>) => void | Promise<void>
type ToolRunStatus = 'queued' | 'executing' | 'completed' | 'failed' | 'cancelled'

interface ToolRunState {
  id: string
  name: string
  arguments: Record<string, unknown>
  status: ToolRunStatus
  concurrencySafe: boolean
  result: ToolResultObj | null
  error: string | null
}

export class ToolExecutionEngine {
  private readonly registry: ToolRegistry
  constructor(registry: ToolRegistry) { this.registry = registry }

  async runBatch(
    toolCalls: ToolCallRequest[],
    opts?: { emit?: StreamEmitter | null; runOne?: (call: ToolCallRequest) => Promise<ToolResultObj>; signal?: AbortSignal | null; maxConcurrency?: number },
  ): Promise<Array<Record<string, unknown>>> {
    const emit = opts?.emit ?? null
    const runOne = opts?.runOne
    const signal = opts?.signal ?? null
    const acquire = makeSemaphore(Math.max(1, Math.trunc(opts?.maxConcurrency ?? DEFAULT_MAX_TOOL_CONCURRENCY)))
    const states = toolCalls.map((call) => this.stateForCall(call))
    if (emit) {
      for (const state of states) {
        await emit(runtimeEvents.toolRunQueued({ id: state.id, name: state.name, arguments: state.arguments }))
      }
    }
    const resultsById = new Map<string, ToolResultObj>()
    let index = 0
    while (index < toolCalls.length) {
      throwIfAborted(signal)
      const state = states[index]!
      if (state.concurrencySafe) {
        const groupCalls: ToolCallRequest[] = []
        const groupStates: ToolRunState[] = []
        while (index < toolCalls.length && states[index]!.concurrencySafe) {
          groupCalls.push(toolCalls[index]!)
          groupStates.push(states[index]!)
          index += 1
        }
        const gathered = await Promise.allSettled(
          groupCalls.map(async (call, i) => {
            const release = await acquire()
            try {
              return await this.runState(call, groupStates[i]!, { emit, runOne, signal })
            } finally {
              release()
            }
          }),
        )
        for (let i = 0; i < groupCalls.length; i++) {
          const call = groupCalls[i]!
          const item = groupStates[i]!
          const raw = gathered[i]!
          if (raw.status === 'rejected') {
            if (raw.reason instanceof TurnPaused) throw raw.reason
            if (raw.reason instanceof CancelledTaskError) throw raw.reason
            const result = ToolResultObj.fromText(`Error: ${raw.reason}`, { isError: true })
            item.status = 'failed'
            item.error = String(raw.reason)
            resultsById.set(call.id, result)
            if (emit) await emit(runtimeEvents.toolRunFailed({ id: call.id, name: call.name, message: String(raw.reason) }))
          } else {
            resultsById.set(call.id, raw.value)
          }
        }
        continue
      }
      resultsById.set(toolCalls[index]!.id, await this.runState(toolCalls[index]!, state, { emit, runOne, signal }))
      index += 1
    }
    return toolCalls.map((call) => ({
      role: 'tool',
      tool_call_id: call.id,
      name: call.name,
      content: (resultsById.get(call.id) ?? ToolResultObj.fromText('')).modelContent,
    }))
  }

  /**
   * 流式工具执行（Wave5）：边流式边入队。canStartEarly 为真的调用（无条件放行、不会暂停）
   * 在 enqueue 时立即起跑；其余调用留到 finish() 按最终响应顺序补跑并对账。
   */
  createStreamingRun(opts: {
    emit?: StreamEmitter | null
    runOne?: (call: ToolCallRequest) => Promise<ToolResultObj>
    signal?: AbortSignal | null
    maxConcurrency?: number
    canStartEarly?: (call: ToolCallRequest) => boolean
  }): {
    enqueue: (call: ToolCallRequest) => void
    finish: (finalCalls: ToolCallRequest[]) => Promise<Array<Record<string, unknown>>>
  } {
    const emit = opts.emit ?? null
    const runOne = opts.runOne
    const signal = opts.signal ?? null
    const canStartEarly = opts.canStartEarly ?? (() => false)
    const acquire = makeSemaphore(Math.max(1, Math.trunc(opts.maxConcurrency ?? DEFAULT_MAX_TOOL_CONCURRENCY)))
    const started = new Map<string, Promise<ToolResultObj>>()
    const queuedEmitted = new Set<string>()

    const emitQueued = async (call: ToolCallRequest): Promise<void> => {
      if (queuedEmitted.has(call.id)) return
      queuedEmitted.add(call.id)
      if (emit) await emit(runtimeEvents.toolRunQueued({ id: call.id, name: call.name, arguments: call.arguments }))
    }

    const runGuarded = (call: ToolCallRequest): Promise<ToolResultObj> => {
      const state = this.stateForCall(call)
      return (async () => {
        const release = await acquire()
        try {
          return await this.runState(call, state, { emit, runOne, signal })
        } finally {
          release()
        }
      })()
    }

    return {
      enqueue: (call: ToolCallRequest): void => {
        void emitQueued(call)
        if (started.has(call.id)) return
        if (signal?.aborted || !canStartEarly(call)) return
        started.set(call.id, runGuarded(call))
      },
      finish: async (finalCalls: ToolCallRequest[]): Promise<Array<Record<string, unknown>>> => {
        throwIfAborted(signal)
        const resultsById = new Map<string, ToolResultObj>()
        for (const call of finalCalls) {
          await emitQueued(call)
          throwIfAborted(signal)
          const early = started.get(call.id)
          // 提前起跑的（只读）并发放行，未起跑的按最终顺序顺序补跑，保留暂停/取消语义
          const result = early ? await early : await runGuarded(call)
          resultsById.set(call.id, result)
        }
        return finalCalls.map((call) => ({
          role: 'tool',
          tool_call_id: call.id,
          name: call.name,
          content: (resultsById.get(call.id) ?? ToolResultObj.fromText('')).modelContent,
        }))
      },
    }
  }

  private stateForCall(call: ToolCallRequest): ToolRunState {
    const tool = this.registry.get(call.name)
    const concurrencySafe = Boolean(tool && tool.isConcurrencySafe(call.arguments))
    return { id: call.id, name: call.name, arguments: call.arguments, status: 'queued', concurrencySafe, result: null, error: null }
  }

  private async runState(
    call: ToolCallRequest,
    state: ToolRunState,
    opts: { emit: StreamEmitter | null; runOne?: (call: ToolCallRequest) => Promise<ToolResultObj>; signal?: AbortSignal | null },
  ): Promise<ToolResultObj> {
    throwIfAborted(opts.signal ?? null)
    state.status = 'executing'
    if (opts.emit) await opts.emit(runtimeEvents.toolRunStarted({ id: state.id, name: state.name }))
    let result: ToolResultObj
    try {
      if (!opts.runOne) {
        result = await this.registry.executeResult(call.name, call.arguments)
      } else {
        result = coerceToolResult(await opts.runOne(call))
      }
    } catch (exc) {
      if (exc instanceof TurnPaused) {
        state.status = 'cancelled'
        if (opts.emit) await opts.emit(runtimeEvents.toolRunCancelled({ id: state.id, name: state.name, reason: 'turn_paused' }))
        throw exc
      }
      if (exc instanceof CancelledTaskError) {
        state.status = 'cancelled'
        if (opts.emit) await opts.emit(runtimeEvents.toolRunCancelled({ id: state.id, name: state.name, reason: 'cancelled' }))
        throw exc
      }
      state.status = 'failed'
      state.error = String(exc)
      if (opts.emit) await opts.emit(runtimeEvents.toolRunFailed({ id: state.id, name: state.name, message: String(exc) }))
      return ToolResultObj.fromText(`Error: ${exc}`, { isError: true })
    }
    throwIfAborted(opts.signal ?? null)
    state.status = result.isError ? 'failed' : 'completed'
    state.result = result
    if (opts.emit) {
      if (result.isError) {
        await opts.emit(runtimeEvents.toolRunFailed({ id: state.id, name: state.name, message: result.summary }))
      } else {
        const output = runtimeEvents.compactRuntimeToolOutput(result.modelContent)
        await opts.emit(
          runtimeEvents.toolRunCompleted({
            id: state.id,
            name: state.name,
            summary: result.summary,
            ...output,
            artifacts: result.artifactPayloads().length ? result.artifactPayloads() : null,
            metadata: Object.keys(result.metadata).length ? result.metadata : null,
          }),
        )
      }
    }
    return result
  }
}

function throwIfAborted(signal: AbortSignal | null | undefined): void {
  if (signal?.aborted) throw new CancelledTaskError('turn')
}

const DEFAULT_MAX_TOOL_CONCURRENCY = 6

/** 手写信号量：并发安全组内节流，防止一次吐出几十个并发工具打满本地 IO。 */
function makeSemaphore(limit: number): () => Promise<() => void> {
  let active = 0
  const waiters: Array<() => void> = []
  const release = (): void => {
    active -= 1
    const next = waiters.shift()
    if (next) next()
  }
  return function acquire(): Promise<() => void> {
    if (active < limit) {
      active += 1
      return Promise.resolve(release)
    }
    return new Promise((resolve) => {
      waiters.push(() => {
        active += 1
        resolve(release)
      })
    })
  }
}

function coerceToolResult(value: ToolResultObj | string): ToolResultObj {
  if (value instanceof ToolResultObj) return value
  const text = String(value)
  return ToolResultObj.fromText(text, { isError: text.startsWith('Error:') })
}
